lib/ssz_ex.ex

```diff
@@ -112,29 +114,26 @@ defmodule LambdaEthereumConsensus.SszEx do
 
   @spec hash_tree_root(list(), {:list, any, non_neg_integer}) ::
           {:ok, Types.root()} | {:error, String.t()}
-  def hash_tree_root(list, {:list, type, size}) do
-    if variable_size?(type) do
-      # TODO
-      # hash_tree_root_list_complex_type(list, {:list, type, size}, limit)
-      {:error, "Not implemented"}
-    else
-      packed_chunks = pack(list, {:list, type, size})
-      limit = chunk_count({:list, type, size})
-      len = length(list)
+  def hash_tree_root(list, {:list, type, _size} = schema) do
+    limit = chunk_count(schema)
+    len = length(list)
+
+    if basic_type?(type) do
+      packed_chunks = pack(list, schema)
       hash_tree_root_list_basic_type(packed_chunks, limit, len)
+    else
+      hash_tree_root_list_composite_type(list, type, limit, len)
     end
   end
 
   @spec hash_tree_root(list(), {:vector, any, non_neg_integer}) ::
           {:ok, Types.root()} | {:error, String.t()}
-  def hash_tree_root(vector, {:vector, type, size}) do
-    if variable_size?(type) do
-      # TODO
-      # hash_tree_root_vector_complex_type(vector, {:vector, type, size}, limit)
-      {:error, "Not implemented"}
-    else
-      packed_chunks = pack(vector, {:list, type, size})
+  def hash_tree_root(vector, {:vector, type, _size} = schema) do
+    if basic_type?(type) do
+      packed_chunks = pack(vector, schema)
       hash_tree_root_vector_basic_type(packed_chunks)
+    else
+      hash_tree_root_vector_composite_type(vector, type)
     end
```

Comment:
We can simplify this by doing:

```elixir
    if basic_type?(type) do
      pack(vector, schema)
    else
      # this is just the `Enum.reduce` in both `hash_tree_root_*_composite_type` functions
      list_hash_tree_roots(vector, schema)
    end
    # we can drop the `_basic_type` here
    |> hash_tree_root_vector_basic_type()
```

Reply:
Yes. Since both are Elixir lists, you should be able to reuse the `list_hash_tree_roots` function. The idea is to move the `Enum.reduce` into that single function, to avoid duplicating code. After that, you can further reduce duplication by hashing lists with the same vector function, but mixing their length at the end (the Rust SSZ lib we use does something similar).

Reply:
I don't understand the second part. Please explain

Reply:
Actually, forget I mentioned that second part. `merkleize_chunks_with_virtual_padding` already does this.